batch/batch/driver/instance_collection/pool.py

```diff
@@ -334,25 +343,50 @@ async def ready_cores_mcpu_from_estimated_job_queue(self):
     LIMIT {share * JOB_QUEUE_SCHEDULING_WINDOW_SECONDS}
   )
 ) AS t
+ORDER BY batch_id ASC, always_run DESC, job_id ASC
 LIMIT {share * JOB_QUEUE_SCHEDULING_WINDOW_SECONDS}
 )
 '''
 
             jobs_query.append(user_job_query)
             jobs_query_args += [user, self.name, user, self.name]
 
-        result = await self.db.select_and_fetchone(
+        result = self.db.select_and_fetchall(
+            # We use the string representation of regions JSON_UNQUOTE for order by instead of the json array
             f'''
-SELECT CAST(COALESCE(SUM(ready_cores_mcpu), 0) AS SIGNED) AS ready_cores_mcpu
+WITH ready_jobs AS (
+    {" UNION ".join(jobs_query)}
```

Comment:
I'm uneasy about an O(N_USERS_RUNNING_JOBS) sized query.

Reply:
Ah, we were already doing this.

Reply:
This was introduced in #12212. I looked at the max query size parameter on GCP and it was 32 MB. I estimated the query for each sample in the union was around 1KB. If your concern is performance, that's a separate issue. But if it's the length of the query, I think we are ok, as I estimated the max query size could handle ~25k simultaneous users.

Reply:
I can double check the `32 MB` number. It's the maximum packet size that MySQL will allow.

Reply:
I couldn't get rid of O(n_samples), but this is what the new query explain plan looks like:

```
mysql> EXPLAIN WITH ready_jobs AS ((
    ->     SELECT user, jobs.batch_id, jobs.job_id, cores_mcpu,
    ->       IF(region IS NULL, NULL, JSON_ARRAYAGG(region)) AS regions,
    ->       COUNT(region) AS n_regions,
    ->       ROW_NUMBER() OVER (ORDER BY batch_id ASC, job_id ASC) DIV 10 AS scheduling_iteration
    ->     FROM jobs FORCE INDEX(jobs_batch_id_state_always_run_cancelled)
    ->     LEFT JOIN batches ON jobs.batch_id = batches.id
    ->     LEFT JOIN batches_cancelled ON batches.id = batches_cancelled.id
    ->     LEFT JOIN job_regions ON jobs.batch_id = job_regions.batch_id AND jobs.job_id = job_regions.job_id
    ->     LEFT JOIN region_ids ON job_regions.region_id = region_ids.region_id    
    ->     WHERE user = 'jigold' AND batches.`state` = 'running' AND jobs.state = 'Ready' AND (always_run = 1 OR always_run = 0 AND batches_cancelled.id IS NULL) AND inst_coll = 'standard'
    ->     GROUP BY jobs.batch_id, jobs.job_id
    ->     ORDER BY jobs.batch_id ASC, jobs.job_id ASC
    -> LIMIT 10)
    -> )
    -> SELECT regions, CAST(COALESCE(SUM(cores_mcpu), 0) AS SIGNED) AS ready_cores_mcpu
    -> FROM (
    ->   SELECT regions, cores_mcpu,
    ->     CAST(ROW_NUMBER() OVER (ORDER BY scheduling_iteration, user, -n_regions DESC, JSON_UNQUOTE(regions), ready_jobs.batch_id, ready_jobs.job_id) AS SIGNED) AS row_num_overall,
    ->     CAST(ROW_NUMBER() OVER (PARTITION BY JSON_UNQUOTE(regions) ORDER BY scheduling_iteration, user, -n_regions DESC, JSON_UNQUOTE(regions), ready_jobs.batch_id, ready_jobs.job_id) AS SIGNED) AS row_num_by_regions
    ->   FROM ready_jobs
    -> ) AS ordered_jobs
    -> GROUP BY regions, row_num_overall - row_num_by_regions
    -> HAVING ready_cores_mcpu > 0
    -> ORDER BY row_num_overall
    -> LIMIT 10;
+----+-------------+-------------------+------------+--------+------------------------------------------+------------------------------------------+---------+--------------------------------------+------+----------+-----------------------------------------------------------+
| id | select_type | table             | partitions | type   | possible_keys                            | key                                      | key_len | ref                                  | rows | filtered | Extra                                                     |
+----+-------------+-------------------+------------+--------+------------------------------------------+------------------------------------------+---------+--------------------------------------+------+----------+-----------------------------------------------------------+
|  1 | PRIMARY     | <derived2>        | NULL       | ALL    | NULL                                     | NULL                                     | NULL    | NULL                                 |   10 |   100.00 | Using temporary; Using filesort                           |
|  2 | DERIVED     | <derived3>        | NULL       | ALL    | NULL                                     | NULL                                     | NULL    | NULL                                 |   10 |   100.00 | Using filesort                                            |
|  3 | DERIVED     | batches           | NULL       | ref    | PRIMARY,batches_user_state               | batches_user_state                       | 424     | const,const                          |    1 |   100.00 | Using where; Using index; Using temporary; Using filesort |
|  3 | DERIVED     | batches_cancelled | NULL       | eq_ref | PRIMARY                                  | PRIMARY                                  | 8       | jigold.batches.id                    |    1 |   100.00 | Using index                                               |
|  3 | DERIVED     | jobs              | NULL       | ref    | jobs_batch_id_state_always_run_cancelled | jobs_batch_id_state_always_run_cancelled | 130     | jigold.batches.id,const              | 2851 |     1.90 | Using index condition; Using where                        |
|  3 | DERIVED     | job_regions       | NULL       | eq_ref | PRIMARY                                  | PRIMARY                                  | 12      | jigold.batches.id,jigold.jobs.job_id |    1 |   100.00 | NULL                                                      |
|  3 | DERIVED     | region_ids        | NULL       | eq_ref | PRIMARY                                  | PRIMARY                                  | 4       | jigold.job_regions.region_id         |    1 |   100.00 | NULL                                                      |
+----+-------------+-------------------+------------+--------+------------------------------------------+------------------------------------------+---------+--------------------------------------+------+----------+-----------------------------------------------------------+
7 rows in set, 3 warnings (0.00 sec)
```